jello/lib.py

```diff
@@ -348,6 +349,46 @@ def create_json(self, data):
             # this code should not run, but just in case something slips by above
             raise TypeError(f'Object is not JSON serializable')
 
+ 
+def format_response(response):
+    """Create schema or JSON/JSON-Lines/Lines"""
+
+    if opts.flatten:
+        it = None
+        if isinstance(response, collections.abc.Iterator):
+            it = response
+        elif isinstance(response, list):
+            it = iter(response)
+        else:
+            raise TypeError('-F/flatten requires the query to return an iterator/generator or list')
```

Comment:
Could we just output the object or scalar here instead?

Reply:
There's no technical reason this can't be done.  In my opinion it suggests an error: the user requested flattening the output and then returned a value that can't be flattened.

There are some surprising results.
"jello -F" whose query returns `1` would output
```
1
```
"jello -F" whose query returns `[1]` would flatten and similarly output
```
1
```

In my mind this is a question of what is most intuitive and least likely to confuse or produce an unexpected result.  For example, I considered letting the user return an _iterable_ in addition to an _iterator_ (so anything providing __iter__() to produce an iterator) but this would lead to confusing situations such as returning the string, whose iterator provides each distinct character (returning "asdf" provides lines with each of "a", "s", "d", and "f").

Similarly, I want to avoid the user trying to stream but accidentally materializing the entire result in memory (they use -S/-F and then start swapping or oom).  To that end maybe allowing the query to return a list is a bad idea and it should always be an iterator (including a generator).

So it's really about what we want to prioritize: try to allow the greatest breadth of return types to just work, or making things as explicit as possible and avoiding expected pitfalls at the expense of a bit more apparent complexity and perhaps being a little less intuitive.

This is also where doing something like https://github.com/kellyjonbrazil/jello/issues/67 splits the difference -- it changes the semantics so that the query is called multiple times, always receiving and returning scalars.  But that shift away from a single call with the entire input being in "_" and the returned value containing the entire output will also confuse.

I'm going back and forth while writing this (which I think you are too).  My opinion at this point is that if it's streaming then the output must be an iterator, including a generator.  Remove even a list.  It'll be tedious for cases where the user wants to return a single value but can be made explicit with a good error message including the recommendation to yield this one value.

But I'm not convinced.  I just want to weight the implication before introducing many options with complex semantics.